### DISCUSSION TITLE
Allowing directory structure and module name mismatch

### DISCUSSION THREAD

**Discussion by yoohaemin** [com-lihaoyi/mill#3953] [Q&A] [answered]
Hi, we have a multi module sbt build in our git repository. 

The root directory contains many unrelated files and folders, and we want to separate the directories that contains sources and the directories that contains other stuffs, so we have all of our module sources inside a root folder called `modules` (`modules/foo`, `modules/bar/baz`, ...).

With sbt, we could determine the root directory of a module in the buildfile, so a `foo` module's root directory can be wherever. In our case, it's `modules/foo`.

Right now I am struggling to have the `build.mill` in the root repository, and not have to have the `module.` prefix in every module (thus many of the cli invocations and in the IDE). Is this possible?

Conceptually, it would be nice to have a way to say: if the module's prefix is `modules`, then drop the prefix. Otherwise, keep the prefix. If they happen to collide, it's ok to blow up. It's our fault.

**Comment by lefou** [comment, marked as answer]
How about this:

```scala
// file: build.mill
package build

import mill._
import mill.scalalib._

object `package` extends RootModule {
  def millSourcePath = super.millSourcePath / "modules"

  object foo extends JavaModule
  object bar extends JavaModule
}
```

```
> mill show __.sources
[1/1] show
{
  "bar.sources": [
    "ref:v0:c984eca8:/tmp/mill-4/modules/bar/src"
  ],
  "foo.sources": [
    "ref:v0:c984eca8:/tmp/mill-4/modules/foo/src"
  ]
}
```

You can always override `millSourcePath`, if your module materializes in a different location.

**Reply by yoohaemin** [reply]
Thank you very much :pray:
